src/common/translation.cpp

```diff
@@ -1452,31 +1438,39 @@ wxString wxTranslations::GetBestTranslation(const wxString& domain,
 wxString wxTranslations::GetBestTranslation(const wxString& domain,
                                             const wxString& msgIdLanguage)
 {
-    wxString lang = GetBestAvailableTranslation(domain);
+    // Determine the best language, including the msgId language, which is always
+    // available because it is present in the code:
+    wxString lang = DoGetBestAvailableTranslation(domain, msgIdLanguage);
+
     if ( lang.empty() )
     {
-        wxArrayString available;
-        available.push_back(msgIdLanguage);
-        available.push_back(msgIdLanguage.BeforeFirst('_'));
-        lang = GetPreferredUILanguage(available);
-        if ( lang.empty() )
-        {
-            wxLogTrace(TRACE_I18N,
-                       "no available language for domain '%s'", domain);
-        }
-        else
-        {
-            wxLogTrace(TRACE_I18N,
-                       "using message ID language '%s' for domain '%s'", lang);
-        }
+        wxLogTrace(TRACE_I18N,
+                   "no available language for domain '%s'", domain);
+    }
+    else if ( lang == msgIdLanguage )
+    {
+        wxLogTrace(TRACE_I18N,
+                   "using message ID language '%s' for domain '%s'", lang);
     }
 
     return lang;
 }
 
 wxString wxTranslations::GetBestAvailableTranslation(const wxString& domain)
 {
-    const wxArrayString available(GetAvailableTranslations(domain));
+    // Determine the best language from the ones with actual translation file:
+    return DoGetBestAvailableTranslation(domain, wxString());
```

Comment:
Passing empty string here makes me wonder if we shouldn't add `msgIdLanguage` parameter to `GetBestAvailableTranslation()` too?

Reply:
IMHO passing an empty string is wrong here, because the message language used in the source code **is** _available_. That is, `msgIdLanguage` should be passed.

The only somewhat problematic use case is when the `msgIdLanguage` of the given domain is **not** _English_ and the standard wxWidgets catalog is **not** available in that language. Then you will have a mix of translation languages.

Reply:
@utelle I agree, and we had this conversation with @vadz in the linked PR before. This part of *this* PR is attempting to preserve the unfortunate `GetBestAvailableTranslation ()` API. If it didn't become part of 3.2 already, I'd argue for removing this function altogether, as it has little if any value — even `GetStdCatalog()`, which motivated all this, can be implemented w/o it just using `GetAvailableTranslations()`.

@vadz What would be the semantics - and usefulness be - then? How would (or even *could*) differ from `GetBestTranslation()` then?

Reply:
Sorry, I got confused (not for the first time...), please disregard my original comment: passing `msgIdLanguage` to this function doesn't make sense because it never uses, or even considers, the original message strings anyhow.

I agree that I wouldn't add `GetBestAvailableTranslation()` today, as the problem using it was supposed to fix is now fixed in a better way. But it's still potentially useful if you really want to find all the catalog files available on the system, for whatever reason, and removing it now would create more problems than it solves.

So, finally, the only thing I'd change would be this:

```suggestion
    // As this function never considers the language of the original messages as being
    // available, pass empty string as message ID language to the helper function.
    return DoGetBestAvailableTranslation(domain, wxString());
```

Reply:
> but it's still potentially useful if you really want to find all the catalog files available on the system

`GetAvailableTranslations()` is better for that...

> for whatever reason, and removing it now would create more problems than it solves.

No argument there :/